### PR TITLE
Improve the question generator

### LINKED ISSUE #92: Improve the question generator with new types of question
# Question generator
Improve the generator by adding new types of questions following Marc's request, this pull request **doesn't aim** to add all the necessary types, but enough to generate a game with different types of questions (5 at least).

### Marc's request 

Système :
A quelle grande classe (= // système1 //) appartient la molécule de // DCI //   ?
-> Les 4 propositions sont des valeurs de la colonne système1.

A quelle grande classe (= // système 2 //) appartient la molécule de // DCI // ?
-> Les 4 propositions sont des valeurs de la colonne système2, ayant le même système1.

Quelle molécule appartient à la classe des // système1 // ?
-> Les 4 propositions sont 4 molécules indépendantes.

Quelle molécule appartient à la classe des // système2 // ?
-> Les 4 propositions sont des molécules ayant le même système1, mais pas le même système2. 

Classe pharmacologique :

A quelle classe pharmacologique  (= // classe1 // )  appartient la molécule de // DCI // ?
-> Les 4 propositions sont des valeurs de la colonne classe1.

A quelle classe pharmacologique  (= // classe2 // )  appartient la molécule de // DCI // ?
-> Les 4 propositions sont des valeurs de la colonne classe2, ayant la même classe1.

A quelle classe pharmacologique  (= // classe3 // )  appartient la molécule de // DCI // ?
-> Les 4 propositions sont des valeurs de la colonne classe3, ayant la même classe2.

Quelle molécule appartient à la classe pharmacologique des // classe1 // ?
-> Les 4 propositions sont des molécules indépendantes.

Quelle molécule appartient à la classe pharmacologique des // classe2 // ?
-> Les 4 propositions sont des molécules ayant la même classe1 mais pas la même classe2.

Quelle molécule appartient à la classe pharmacologique des // classe3 // ?
-> Les 4 propositions sont des molécules ayant la même classe2 mais pas la même classe3.

Indications:
Quelle est l'indication de la molécule de // DCI // ?
Quelle molécule a pour indication : // indication // ?

Effets indésirables :
Quelle molécule a pour effet indésirable notable : // effet indésirable // ?
Quel effet indésirable notable peut se produire avec la molécule de // DCI // ?

Interactions:
La molécule de // DCI // est // interaction // ?
Quelle molécule est // interaction // ?

Structures chimiques:
Quelle est la structure de la molécule de // DCI // ?
A quelle DCI correspond la structure suivante ?
A quelle classe pharmacologique appartient la molécule de structure suivante ? (mêmes règles que pour les classes pharmacologiques, en suivant la hiérarchie des classes )

### PR DESCRIPTION
**Pull request by FrancoisPog**
## Changes
closes #92 

- [x] Create sql scripts to fetch questions data
- [x] Format the question in the `question` controller
- [x] Update [doc](https://github.com/nathanaelhoun/glowing-octo-guacamole/wiki/Question-generator)
- [x] Adapt the client

### Not enough data
If the server cannot generate a question because the database does not contain enough data, the 422 error is sent. In the client's training view, if a 422 error is received, it retry a request (max 10 times).

### Re : Note about tests
Instead of inserting data in the `before` and deleting it in the `after`, it's more convenient to delete data and then insert the required data in the `before`.

### CODE REVIEW

**Review by valentinperignon — APPROVED** [review]
LGTM!

File: server/test/question.test.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  for (let type = 1; type <= 10; ++type) {
+    it("Question type " + type, async () => {
+      const res = await requestAPI("question/" + type);
+      expect(res.status).equals(422);
+    });
+  }
+});
+
+describe("Question generation", function () {
+  const questionTypes = [1, 2, 3, 5, 6, 7, 8, 9];
```
Comment (nathanaelhoun) [inline server/test/question.test.js:27]: Why not also `10`?

**Review by nathanaelhoun — COMMENTED** [review]
Some non-blocking comments

File: server/test/question.test.js (hunk 1 above)
Comment (FrancoisPog) [reply server/test/question.test.js:27]: With the current data we can't generate question of type 4 and 10.

File: server/modules/data_importer/data_importer.js [hunk 2]
```diff
@@ -37,7 +37,7 @@ export function createSqlToInsertAllData(data) {
     script += createSqlToInsertProperty(property, data[property]);
   }
   script += createSqlToInsertAllMolecules(data.molecules);
-
+  //console.error(script);
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/modules/data_importer/data_importer.js:40]: yep

File: server/test/question.test.js (hunk 1 above)
Comment (nathanaelhoun) [reply server/test/question.test.js:27]: Ok, I think a comment could be useful to clarify this 😉

File: server/controllers/question.js [hunk 3]
```diff
@@ -48,51 +101,98 @@ export default { generateQuestion };
...
+
+/**
+ * Format a question from the SQL reponse data
+ * @param {object[]} data The sql response data
+ * @returns {{subject : string, goodAnswer : number, answers : string[]}} The formatted question
+ */
+function formatQuestion(data) {
+  const badAnswers = data.map((e) => e.bad_answer);
+  const goodAnswer = data[0].good_answer;
+  const randomIndex = Math.floor(Math.random() * 4);
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/controllers/question.js:175]: I'll change that !

File: server/controllers/question.js [hunk 4]
```diff
@@ -48,51 +101,98 @@ export default { generateQuestion };
...
+class NotEnoughDataError extends Error {
+  constructor() {
+    super();
+    this.message = "This type of question is currently not available";
+    this.name = "Not Enough Data";
+    this.code = "NED";
+  }
+
+  static isInstance(error) {
+    return error instanceof NotEnoughDataError;
```
[the remark this answers is not in the document]
Comment (FrancoisPog) [reply server/controllers/question.js:196]: there is no technical reason, but I've read  somewhere that using functions is more expressive than built-in operators, but it doesn't really matter 😅

**Pull request merged**